app/Helpers/Storage.php

```diff
@@ -52,4 +55,58 @@ public function uploadS3($presigned_url)
         ];
     }
 
+
+    /**
+     * [uploadFTP description]
+     * @param  array $FTPcredentials The credentials for the FTP server
+     * @return array                 Upload's result
+     */
+    public function uploadFTP($FTPcredentials, $invoiceId) {
+        if (ends_with($FTPcredentials["path"], "/") == false) {
+            $FTPcredentials["path"] = $FTPcredentials["path"] . "/";
+        }
+
+        $adapter = new Adapter([
+            'host' => $FTPcredentials["host"],
+            'username' => $FTPcredentials["username"],
+            'password' => $FTPcredentials["password"],
+            'port' => isset($FTPcredentials["port"]) ? $FTPcredentials["port"] : 21,
+            'root' => $FTPcredentials["path"],
+            'passive' => isset($FTPcredentials["passive"]) ? $FTPcredentials["passive"] : true,
+            'ssl' => isset($FTPcredentials["ssl"]) ? $FTPcredentials["ssl"] : false,
+            'timeout' => 300
+        ]);
+
+        try {
+            $adapter->getConnection();
+        } catch (\RuntimeException $e) {
+            return [
+                'uploaded' => false,
+                'message' => $e->getMessage(),
+                'path' => null,
+            ];
+        }
+
+        $filesystem = new Filesystem($adapter);
+
+        $filePath = "invoice_". $invoiceId ."_" . time() .".pdf";
+
+        $response = $filesystem->write($filePath , $this->doc, ['visibility' => 'public']);
+        if (!$response) {
+            return [
+                'uploaded' => false,
+                'message' => "The invoice could not be uploaded to the FTP server.",
+                'path' => null
+            ];
+        }
+
+        return [
+            'uploaded' => true,
+            'message' => null,
+            'path' => $FTPcredentials["path"] . $filePath
```

Comment:
It should start with `ftp://<host>:<port>/.......`

Reply:
I am thinking what if the public path is different than the relative path on the server where it was actually uploaded? Maybe to have a different, second parameter for a public path? and the filename to be concatenated with that public path?

Reply:
Invoices should never be available publicly.

But you're right, if another service want to access to this file through a NFS mounted directory, returning only the relative path would be enough here.